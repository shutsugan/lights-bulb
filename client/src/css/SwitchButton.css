.SwitchButton {
  --uiSwitchSize: var(--switchSize, 64px);
  --uiSwitchBgColor: var(--switchBgColor, #ddd);
  --uiSwitchBgColorActive: var(--switchBgColorActive, #4ed164);
  --uiSwitchBorderColorActive: var(--switchBorderColorActive, #fff);
  --uiSwitchBorderColorFocus: var(--switchBgColorFocus, #43b556);
  --uiSwitchButtonBgColor: var(--switchButtonBgColor, #fff);  
  
  display: inline-block;
  position: relative;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
}

.SwitchButton__label {
  display: block;
  width: 100%;
  height: 100%;


  font-size: .95em;
  text-align: left;
  color: #FFF;
  padding: 0 3px;
}

.SwitchButton__toggle {
  width: 0;
  height: 0;
  opacity: 0;
  
  position: absolute;
  top: 0;
  left: 0;
}

.SwitchButton__toggle:focus ~ .SwitchButton__label {
  box-shadow: 0 0 0 var(--uiSwitchThickFocus, 1.5px) var(--uiSwitchBorderColorFocus);
}

.SwitchButton__toggle:checked:focus ~ .SwitchButton__label {
  box-shadow: 0 0 0 var(--uiSwitchThickFocus, 1.5px) var(--uiSwitchBorderColorActive);
}

.SwitchButton__label:before, .SwitchButton__label:after {
  content: "";
  cursor: pointer;

  position: absolute;
  top: 0;
  left: 0;
}

.SwitchButton__label:before {
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  background-color: var(--uiSwitchBgColor);
}

.SwitchButton__label:after{
  top: 50%;
  z-index: 3;
  transition: transform .4s cubic-bezier(0.44,-0.12, 0.07, 1.15);
}

.SwitchButton {
  --uiSwitchIndent: var(--switchIndent, 2px);
  --uiSwitchBorderRadius: var(--switchBorderRadius, 60px);

  width: var(--uiSwitchSize);
  height: calc((var(--uiSwitchSize) / 2));
  border-radius: var(--uiSwitchBorderRadius);    
  background-color: var(--uiSwitchBgColorActive);
}

.SwitchButton__label {
  border-radius: var(--uiSwitchBorderRadius);    
}

.SwitchButton__label:before {
  border-radius: var(--uiSwitchBorderRadius);    
  transition: opacity .2s ease-out .1s, transform .2s ease-out .1s;
  transform: scale(1);
  opacity: 1;  
}

.SwitchButton__toggle:checked ~ .SwitchButton__label:before {
  transform: scale(0);
  opacity: .7;
}

.SwitchButton__toggle ~ .SwitchButton__label:after {
  width: calc((var(--uiSwitchSize) / 2) - calc(var(--uiSwitchIndent) * 2));
  height: calc((var(--uiSwitchSize) / 2) - calc(var(--uiSwitchIndent) * 2));
  transform: translate3d(var(--uiSwitchIndent), -50%, 0);

  background-color: var(--uiSwitchButtonBgColor);
  border-radius: 100%;
}

.SwitchButton__toggle:checked ~ .SwitchButton__label:after {
  transform: translate3d(calc(var(--uiSwitchSize) - calc((var(--uiSwitchSize) / 2) - calc(var(--uiSwitchIndent) * 2)) - var(--uiSwitchIndent)), -50%, 0);
}

.SwitchButton {
  --switchSize: 32px;
  margin-top: 0px;
  opacity: .4;
  pointer-events: none;
}

.SwitchButton {
  --switchBgColorActive: #4cb312;
  --switchBgColorFocus: #317d06;
}

.active {
  opacity: 1;
  pointer-events: auto;
}

a{
  text-decoration: none;
  color: #222;
}