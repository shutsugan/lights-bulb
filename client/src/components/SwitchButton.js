import React, {Component} from 'react';

import '../css/SwitchButton.css';

class SwitchButton extends Component {
	constructor(props) {
		super(props);

		this.state = {
			state: this.props.state
		}
	}

	/**
 	 * Toggle the state of the switch.
 	 * @param {Event} event
	 */
	_toggle = event => {
		this.props._toggle(!this.state.state);
		this._update(event);
		this.setState({state: !this.state.state})
	}

	/**
 	 * Update the state of the current light bubl.
 	 * @param {Event} event
	 */
	_update = event => {
		const id = event.target.id;

		const active = this.state.state;
		const brightness = (!active) ? 100 : 0;
		const data = {
			active: !active,
			brightness: brightness
		}

		this.props._update(id, data);
	}


	render() {
		const active = this.props.selected === this.props.id ? 'active' : '';

		return(
			<label className={`SwitchButton ${active}`}>
	        	<input 
	        		type="checkbox"
	        		checked={this.props.state} 
	        		id={this.props.id}
	        		className="SwitchButton__toggle" 
	        		onChange={this._toggle}
	        		/>
	        	<span className="SwitchButton__label">on</span>
	        </label> 
		);
	}
}

export default SwitchButton;