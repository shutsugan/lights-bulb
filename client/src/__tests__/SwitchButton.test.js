import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import SwitchButton from '../components/SwitchButton';

const device = {
    id: 3,
    active: false
}

const device_2 = {
	id: 1,
	active: true
}

describe('SwitchButton', () => {
    it('should be defined', () => {
      expect(SwitchButton).toBeDefined();
    });
  
    it('should render correctly', () => {
      const tree = shallow(<SwitchButton id={device.id} state={device.active} />);
      expect(tree).toMatchSnapshot();
    });

    it('should match the device state', () => {
      const tree = shallow(<SwitchButton id={device.id} state={device.active} />);
      const tree_2 = shallow(<SwitchButton id={device.id} state={device_2.active} />);

      expect(tree.unrendered.props.state).toEqual(false);
      expect(tree_2.unrendered.props.state).toEqual(true);
    });
});