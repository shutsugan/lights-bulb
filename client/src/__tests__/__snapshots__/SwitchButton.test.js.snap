// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchButton should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <label
    className="SwitchButton "
>
    <input
        checked={false}
        className="SwitchButton__toggle"
        id={3}
        onChange={[Function]}
        type="checkbox"
    />
    <span
        className="SwitchButton__label"
    >
        on
    </span>
</label>,
  "nodes": Array [
    <label
      className="SwitchButton "
>
      <input
            checked={false}
            className="SwitchButton__toggle"
            id={3}
            onChange={[Function]}
            type="checkbox"
      />
      <span
            className="SwitchButton__label"
      >
            on
      </span>
</label>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_context": Object {},
    "_element": <SwitchButton
      id={3}
      state={false}
/>,
    "_instance": SwitchButton {
      "_toggle": [Function],
      "_update": [Function],
      "context": Object {},
      "props": Object {
        "id": 3,
        "state": false,
      },
      "refs": Object {},
      "state": Object {
        "state": false,
      },
      "updater": Updater {
        "_renderer": [Circular],
      },
    },
    "_newState": null,
    "_rendered": <label
      className="SwitchButton "
>
      <input
            checked={false}
            className="SwitchButton__toggle"
            id={3}
            onChange={[Function]}
            type="checkbox"
      />
      <span
            className="SwitchButton__label"
      >
            on
      </span>
</label>,
    "_rendering": false,
    "_updater": Updater {
      "_renderer": [Circular],
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SwitchButton
    id={3}
    state={false}
/>,
}
`;
